File: Extensions/fishcatcher/src/engine/formaction.js
// S20: form-action destination. Pure: probe-derived hostnames in, reasons out.
// A login form that posts somewhere other than the page's own site is the
// classic credential-harvest tell. Bundled into background.js by build.mjs;
// every top-level name is formAction-/FORM_ACTION-prefixed for the bundler.
import { registrableDomain } from './psl.js';
import { isIpAddress } from './signals.js';

export const FORM_ACTION_WEIGHT = 40;       // cross-domain post → high alone
export const FORM_ACTION_EXFIL_WEIGHT = 60; // known credential-exfil sink

// Hosts (and their subdomains) scammers use to collect what a victim types.
const FORM_ACTION_EXFIL_HOSTS = [
  'api.telegram.org', 'discord.com', 'discordapp.com', 'formspree.io', 'getform.io',
  'formsubmit.co', 'formcarry.com', 'docs.google.com', 'script.google.com',
  'webhook.site', 'pipedream.net', 'ngrok.io', 'ngrok-free.app', 'ngrok.app'
];

function formActionIsExfil(host) {
  if (isIpAddress(host)) return true;
  if (/(^|\.)requestbin\./.test(host)) return true;
  return FORM_ACTION_EXFIL_HOSTS.some((d) => host === d || host.endsWith('.' + d));
}

// ctx = { registrable, knownLegit, aitm }; aitm.formActions = [hostname].
// Known-legit (top-100k) sites post to SSO hosts cross-domain routinely → silent.
export function runFormAction(ctx, data) {
  const out = { score: 0, reasons: [] };
  const hosts = ctx.aitm?.formActions;
  if (ctx.knownLegit || !Array.isArray(hosts) || !hosts.length) return out;
  let cross = null;
  for (const raw of hosts) {
    const host = String(raw).toLowerCase();
    const dest = isIpAddress(host) ? host : registrableDomain(host, data.psl);
    if (dest === ctx.registrable) continue;
    if (formActionIsExfil(host)) {
      out.score = FORM_ACTION_EXFIL_WEIGHT;
      out.reasons.push({ key: 'reasonFormExfil', params: [dest], weight: FORM_ACTION_EXFIL_WEIGHT });
      return out;
    }
    if (data.safeList?.has(dest) || data.aitmIdp?.has(dest) || data.aitmMediation?.has(dest)) continue;
    // Same-brand SSO (accounts.brand.com ↔ brand-mail.com) is not a mismatch.
    if (data.brands.some((b) => b.domains.includes(ctx.registrable) && b.domains.includes(dest))) continue;
    cross ??= dest;
  }
  if (cross) {
    out.score = FORM_ACTION_WEIGHT;
    out.reasons.push({ key: 'reasonFormAction', params: [cross], weight: FORM_ACTION_WEIGHT });
  }
  return out;
}

File: Search.Kit/FishCatcher/FormAction.cs
namespace SearchKit.FishCatcher;

/// S20, where a sign-in form sends what you type. Port of formaction.js: a
/// password form that posts to another site is the classic harvesting tell,
/// and one that posts to a known collection service is worse.
public static class FormAction
{
    public const int Weight = 40;      // cross-domain post: high alone
    public const int ExfilWeight = 60; // a known credential-collection sink

    // Hosts (and their subdomains) scammers use to collect what a victim types.
    private static readonly string[] ExfilHosts =
    [
        "api.telegram.org", "discord.com", "discordapp.com", "formspree.io", "getform.io",
        "formsubmit.co", "formcarry.com", "docs.google.com", "script.google.com",
        "webhook.site", "pipedream.net", "ngrok.io", "ngrok-free.app", "ngrok.app",
    ];

    public static bool IsExfil(string host)
    {
        if (Signals.IsIpAddress(host)) return true;
        // /(^|\.)requestbin\./
        int at = host.IndexOf("requestbin.", StringComparison.Ordinal);
        while (at >= 0)
        {
            if (at == 0 || host[at - 1] == '.') return true;
            at = host.IndexOf("requestbin.", at + 1, StringComparison.Ordinal);
        }
        return ExfilHosts.Any(d => host == d || host.EndsWith("." + d, StringComparison.Ordinal));
    }

    /// Known-legitimate sites post to SSO hosts across domains all the time, so they're never flagged.
    internal static int Run(string registrable, bool knownLegit, AitmFacts aitm, FishData data, List<Signal> reasons)
    {
        if (knownLegit || aitm.FormActions.Count == 0) return 0;
        string? cross = null;
        foreach (var raw in aitm.FormActions)
        {
            var host = JsText.Lower(raw);
            // The destination is named on the warning, and the page supplied
            // it: anything that isn't a host name is the page's own text.
            if (!PageFacts.IsHost(host)) continue;
            var dest = Signals.IsIpAddress(host) ? host : data.Psl.RegistrableDomain(host);
            if (dest == registrable) continue;
            if (IsExfil(host))
            {
                reasons.Add(new Signal("reasonFormExfil", ExfilWeight, dest));
                return ExfilWeight;
            }
            if (data.SafeList.Contains(dest) || data.AitmIdp.Contains(dest) || data.AitmMediation.Contains(dest)) continue;
            // Same-brand SSO (accounts.brand.com and brand-mail.com) is not a mismatch.
            if (data.Brands.Any(b => b.Domains.Contains(registrable) && b.Domains.Contains(dest))) continue;
            cross ??= dest;
        }
        if (cross is null) return 0;
        reasons.Add(new Signal("reasonFormAction", Weight, cross));
        return Weight;
    }
}
